### PR TITLE
simplex: step-6

### PR DESCRIPTION
**Pull request by marcfehling**
~Blocked by #11590 and #11637.~ (EDIT: Have been merged.)

Port of `step-6` for triangles.

@peterrum -- FYI

### CODE REVIEW

**Comment by marcfehling** [comment]
step-6

File: tests/simplex/step-06.cc [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+// Step-06 on a simplex mesh.
+//
+// The following changes had to be made due to incompatibilities:
+//  - Manifolds are disabled
+//    All manifold ids from the triangulation object generated with
+//    GridGenerator::hyper_ball() had to be removed
```
Comment (peterrum) [inline tests/simplex/step-06.cc:21]: Why? Could you do it like this: https://github.com/dealii/dealii/blob/b59c748f39b00cf111ab9eec96c115d8aa9792a2/tests/simplex/step-01.cc#L65-L67

**Review by peterrum — COMMENTED** [review]
Excellent!

File: tests/simplex/step-06.cc [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+//  - Error estimation has to be skipped
+//    KellyErrorEstimator requires QProjector::project_to_all_subfaces()
```
Comment (luca-heltai) [inline tests/simplex/step-06.cc:23]: but below it looks like you are actually calling it....

**Comment by luca-heltai** [comment]
This seems to fail:
~~~
simplex/step-06.debug: RUN failed. ------ Additional output on stdout/stderr:


--------------------------------------------------------
An error occurred in line <889> of file </home/runner/work/dealii/dealii/source/fe/fe.cc> in function
    const dealii::FullMatrix<double>& dealii::FiniteElement<<anonymous>, <anonymous> >::constraints(const dealii::internal::SubfaceCase<dim>&) const [with int dim = 2; int spacedim = 2]
The violated condition was: 
    (this->n_dofs_per_face(face_no) == 0) || (interface_constraints.m() != 0)
Additional information: 
    The finite element for which you try to obtain hanging node constraints does not appear to implement them.
--------------------------------------------------------

Aborted (core dumped)
CMakeFiles/simplex.step-06.debug.diff.dir/build.make:88: recipe for target 'step-06.debug/output' failed
CMakeFiles/Makefile2:709: recipe for target 'CMakeFiles/simplex.step-06.debug.diff.dir/all' failed
CMakeFiles/Makefile2:716: recipe for target 'CMakeFiles/simplex.step-06.debug.diff.dir/rule' failed
Makefile:382: recipe for target 'simplex.step-06.debug.diff' failed
~~~

File: tests/simplex/step-06.cc (hunk 1 above)
Comment (marcfehling) [reply tests/simplex/step-06.cc:21]: Will give it a try!

**Comment by marcfehling** [comment]
> This seems to fail:
> 
> ```
> [...]
> ```

Will work once blocking PRs are merged.

**Comment by bangerth** [comment]
It fails with this error:
```

simplex/step-06.debug: RUN failed. ------ Additional output on stdout/stderr:


--------------------------------------------------------
An error occurred in line <2283> of file </home/runner/work/dealii/dealii/source/fe/mapping_fe.cc> in function
    std::vector<dealii::Point<spacedim> > dealii::MappingFE<dim, spacedim>::compute_mapping_support_points(const typename dealii::Triangulation<dim, spacedim>::cell_iterator&) const [with int dim = 2; int spacedim = 2; typename dealii::Triangulation<dim, spacedim>::cell_iterator = dealii::TriaIterator<dealii::CellAccessor<2, 2> >]
The violated condition was: 
    check_all_manifold_ids_identical(cell)
Additional information: 
    All entities of a cell need to have the same boundary id as the cell has.
--------------------------------------------------------
```
Can you investigate?

**Comment by luca-heltai** [comment]
> It fails with this error:
> 
> ```
> 
> simplex/step-06.debug: RUN failed. ------ Additional output on stdout/stderr:
> 
> 
> --------------------------------------------------------
> An error occurred in line <2283> of file </home/runner/work/dealii/dealii/source/fe/mapping_fe.cc> in function
>     std::vector<dealii::Point<spacedim> > dealii::MappingFE<dim, spacedim>::compute_mapping_support_points(const typename dealii::Triangulation<dim, spacedim>::cell_iterator&) const [with int dim = 2; int spacedim = 2; typename dealii::Triangulation<dim, spacedim>::cell_iterator = dealii::TriaIterator<dealii::CellAccessor<2, 2> >]
> The violated condition was: 
>     check_all_manifold_ids_identical(cell)
> Additional information: 
>     All entities of a cell need to have the same boundary id as the cell has.
> --------------------------------------------------------
> ```
> 
> Can you investigate?

I think this is due to the fact that currently high order mapping only works for uniform manifold ids (i.e., either all flats, or all curved with the same ids). The current implementation would work for hyper shells (and simplices) but not for balls (where flat manifold / transfinite manifold is used on cell at boundaries, and polar manifold is used on outer cells).

Reset all manifold ids to flat. 

We'll address this at some point. BTW, @peterrum , I guess the error message should have been "All entities of a cell need to have the same manifold id as the cell has.", not "boundary".

**Comment by marcfehling** [comment]
> Reset all manifold ids to flat.

@luca-heltai I'll remove the last commit then. (I didn't test it yet)

**Comment by bangerth** [comment]
Do I understand correctly that you want to let the test stand as is?

**Comment by marcfehling** [comment]
> Do I understand correctly that you want to let the test stand as is?

Yes, I would like to leave it like that for now. Like the other simplex tests, we point out in the preamble what is currently not working. We should address these issues in the future (as @luca-heltai also pointed out). This tests represents what we are currently capable of.

**Review by bangerth — APPROVED** [review]

**Pull request merged**